be/src/exec/join_hash_map.tpp

```diff
@@ -808,63 +811,77 @@ void JoinHashMap<LT, BuildFunc, ProbeFunc>::_search_ht_remain(RuntimeState* stat
     _probe_state->count = match_count;
 }
 
+#define DO_PROBE(X, Y)                                                                  \
+    if (enable_interleaving) {                                                          \
+        if constexpr (first_probe) {                                                    \
+            auto group_size = std::abs(state->query_options().interleaving_group_size); \
+            _probe_state->cur_probe_index = 0;                                          \
+            _probe_state->handles.clear();                                              \
+            for (int i = 0; i < group_size; ++i) {                                      \
+                _probe_state->handles.insert(X(state, build_data, data));               \
+            }                                                                           \
+            _probe_state->active_coroutines = group_size;                               \
+        }                                                                               \
+        _probe_coroutine<first_probe, Y>(state, build_data, data);                      \
+    } else {                                                                            \
+        X<first_probe>(state, build_data, data);                                        \
+    }
+
 template <LogicalType LT, class BuildFunc, class ProbeFunc>
 template <bool first_probe>
 void JoinHashMap<LT, BuildFunc, ProbeFunc>::_search_ht_impl(RuntimeState* state, const Buffer<CppType>& build_data,
                                                             const Buffer<CppType>& data) {
+    bool enable_interleaving = state->query_options().interleaving_group_size < 0; // force enable
+    if (state->query_options().interleaving_group_size > 0) {
+        enable_interleaving = ((_table_items->probe_bytes > (1UL << 25) && _table_items->keys_per_bucket > 1.5) ||
```

Comment:
why check _table_items->keys_per_bucket > 1 ?

Reply:
as tested, if keys_per_bucket is small, the benefit is not obvious, so limit it to get obvious benefit. if keys_per_bucket = 1, each valid bucket just has one key, no need to access next, less random access.